Tools/ci-run.sh

```diff
@@ -3,125 +3,164 @@
 else
   python -m pip install -U pip setuptools wheel || exit 1
 
-  if [ -n "${PYTHON_VERSION##*-dev}" ]; then
+  if [[ $PYTHON_VERSION == *"-dev" || $COVERAGE == "1" ]]; then
     python -m pip install -r test-requirements.txt || exit 1
 
-    if [ "${PYTHON_VERSION##pypy*}" -a "${PYTHON_VERSION##3.[4789]*}" ]; then
+    if [[ $PYTHON_VERSION == "pypy"* && $PYTHON_VERSION == "3."[4789]* ]]; then
       python -m pip install -r test-requirements-cpython.txt || exit 1
     fi
   fi
 fi
 
-if [ "$TEST_CODE_STYLE" == "1" ]; then
-  STYLE_ARGS="--no-unit --no-doctest --no-file --no-pyregr --no-examples";
+if [[ $TEST_CODE_STYLE == "1" ]]; then
+  STYLE_ARGS="--no-unit --no-doctest --no-file --no-pyregr --no-examples"
   python -m pip install -r doc-requirements.txt || exit 1
 else
-  STYLE_ARGS="--no-code-style";
+  STYLE_ARGS="--no-code-style"
 
   # Install more requirements
-  if [ -n "${PYTHON_VERSION##*-dev}" ]; then
-    if [ -z "${BACKEND##*cpp*}" ]; then
+  if [[ $PYTHON_VERSION == *"-dev" ]]; then
+    if [[ $BACKEND == *"cpp"* ]]; then
       echo "WARNING: Currently not installing pythran due to compatibility issues"
       # python -m pip install pythran==0.9.5 || exit 1
-    fi
-
-    if [ "$BACKEND" != "cpp" -a -n "${PYTHON_VERSION##pypy*}" -a -n "${PYTHON_VERSION##2*}" -a -n "${PYTHON_VERSION##*3.4}" ]; then
+    elif [[ $PYTHON_VERSION == "pypy"* && $PYTHON_VERSION == "2"* && $PYTHON_VERSION == *"3.4" ]]; then
       python -m pip install mypy || exit 1
     fi
   fi
 fi
 
 # Run tests
+echo "==== Running tests ===="
 ccache -s 2>/dev/null || true
 export PATH="/usr/lib/ccache:$PATH"
 
-if [ "$NO_CYTHON_COMPILE" != "1" -a -n "${PYTHON_VERSION##pypy*}" ]; then
-  CFLAGS="-O2 -ggdb -Wall -Wextra $(python -c 'import sys; print("-fno-strict-aliasing" if sys.version_info[0] == 2 else "")')" \
-  python setup.py build_ext -i \
-          $(if [ "$COVERAGE" == "1" ]; then echo " --cython-coverage"; fi) \
-          $(if [ "$CYTHON_COMPILE_ALL" == "1" ]; then echo " --cython-compile-all"; fi) \
-          $(python -c 'import sys; print("-j5" if sys.version_info >= (3,5) else "")') \
-      || exit 1
-  if [ -z "$COVERAGE" -a -z "$STACKLESS" -a -z "$LIMITED_API" -a -z "$CYTHON_COMPILE_ALL" -a -z "$EXTRA_CFLAGS" -a -n "${BACKEND//*cpp*}" ]; then
+if [[ "$OSTYPE" == "msys" ]]; then  # for MSVC cl
+  # /wd will disable warning
+  # 4711 warns that function `x` selected for automatic inline expansion
+  # 4127 warns that conditional expression is constant, should be fixed here https://github.com/cython/cython/pull/4317
+  # (off by default) 5045 warns that Compiler will insert Spectre mitigation for memory load if /Qspectre switch specified
+  # (off by default) 4820 warns about the code in Python\3.9.6\x64\include ...
+  WARNARGS="/W4 /wd4711 /wd4127 /wd5045 /wd4820"
+  DEBUG_INFO="/Z7"
+  NO_OPTIMIZATION="-Od"
+else
+  WARNARGS="-Wall -Wextra"
+  DEBUG_INFO="-ggdb"
+  NO_OPTIMIZATION="-O0"
```

Comment:
I don't think we need different variables for them. It just makes it less obvious how they will be used. All of these are just `CFLAGS`. Put them there.

Reply:
There is a place where we use WARNARGS + DEBUG_INFO with -O3, so ... if I join them with cflags it will cause unnecessary repetition, but I will try to work around

Reply:
here are all the use cases:

`CFLAGS="-O2 $DEBUG_INFO $WARNARGS $ALIASING"`
`CFLAGS="$NO_OPTIMIZATION $DEBUG_INFO" $PYTHON_DBG`
`CFLAGS="$NO_OPTIMIZATION $DEBUG_INFO $WARNARGS $EXTRA_CFLAGS"`

I do not see a better solution than adding `DEBUG_INFO` to `CFLAGS`, and then we will have to write like this:

`CFLAGS="$CFLAGS -O2 $WARNARGS $ALIASING"`
`CFLAGS="$CFLAGS $NO_OPTIMIZATION" $PYTHON_DBG`
`CFLAGS="$CFLAGS $NO_OPTIMIZATION $WARNARGS $EXTRA_CFLAGS"`

Not much better ..

Reply:
> here are all the use cases:
> 
> `CFLAGS="-O2 $DEBUG_INFO $WARNARGS $ALIASING"`
> `CFLAGS="$NO_OPTIMIZATION $DEBUG_INFO" $PYTHON_DBG`
> `CFLAGS="$NO_OPTIMIZATION $DEBUG_INFO $WARNARGS $EXTRA_CFLAGS"`
> 
> I do not see a better solution than adding `DEBUG_INFO` to `CFLAGS`, and then we will have to write like this:
> 
> `CFLAGS="$CFLAGS -O2 $WARNARGS $ALIASING"`
> `CFLAGS="$CFLAGS $NO_OPTIMIZATION" $PYTHON_DBG`
> `CFLAGS="$CFLAGS $NO_OPTIMIZATION $WARNARGS $EXTRA_CFLAGS"`
> 
> Not much better ..

I think it's time to clean this up.

We can always add `$WARNARGS` and `$DEBUG_INFO` globally to `CFLAGS` since it doesn't hurt.

`-O2` can go to the end, then it's just `CFLAGS="$CFLAGS -O2" python setup.py …`. We can then add `-O0` or `-Og` to `CFLAGS` by default.

AFAICT, that leaves two cases where we add a single additional option. I think that's more than bearable.

Reply:
@scoder Are you sure that the argument (i.e. `-O2`) coming after the current (i.e. `-O0`) overwrite it?

Reply:
[partial fix](https://github.com/cython/cython/pull/4324/commits/38c80151ef87a19b874b77975797d71b9af7b10d)

Reply:
@scoder Are you ok with the current implementation? If not, then we need to make sure that we replace the default optimization argument with a new one when we ascribe it to the end.

Reply:
[resolved](https://github.com/cython/cython/pull/4324/commits/78b22fc3da2ec23e4911ed6f3b50f44cb5c15b8d)